server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java

```diff
@@ -0,0 +1,180 @@
+
+public class SynonymsManagementAPIService {
+    public static final String SYNONYMS_INDEX_NAME_PATTERN = ".synonyms-*";
+    public static final String SYNONYMS_INDEX_CONCRETE_NAME = ".synonyms-1";
+    public static final String SYNONYMS_ALIAS_NAME = ".synonyms";
+
+    public static final String SYNONYMS_FEATURE_NAME = "synonyms";
+    public static final String SYNONYMS_SET_FIELD = "synonyms_set";
+    public static final String SYNONYMS_FIELD = "synonyms";
+
+    private final Client client;
+
+    public static final String SYNONYMS_ORIGIN = "synonyms";
+    public static final SystemIndexDescriptor SYNONYMS_DESCRIPTOR = SystemIndexDescriptor.builder()
+        .setIndexPattern(SYNONYMS_INDEX_NAME_PATTERN)
+        .setDescription("Synonyms index for synonyms managed through APIs")
+        .setPrimaryIndex(SYNONYMS_INDEX_CONCRETE_NAME)
+        .setAliasName(SYNONYMS_ALIAS_NAME)
+        .setMappings(mappings())
+        .setSettings(settings())
+        .setVersionMetaKey("version")
+        .setOrigin(SYNONYMS_ORIGIN)
+        .build();
+
+    public SynonymsManagementAPIService(Client client) {
+        this.client = new OriginSettingClient(client, SYNONYMS_ORIGIN);
+    }
+
+    private static XContentBuilder mappings() {
+        try {
+            XContentBuilder builder = jsonBuilder();
+            builder.startObject();
+            {
+                builder.startObject(SINGLE_MAPPING_NAME);
+                {
+                    builder.startObject("_meta");
+                    {
+                        builder.field("version", Version.CURRENT.toString());
+                    }
+                    builder.endObject();
+                    builder.field("dynamic", "strict");
+                    builder.startObject("properties");
+                    {
+                        builder.startObject(SYNONYMS_FIELD);
+                        {
+                            builder.field("type", "match_only_text");
+                        }
+                        builder.endObject();
+                        builder.startObject(SYNONYMS_SET_FIELD);
+                        {
+                            builder.field("type", "keyword");
+                        }
+                        builder.endObject();
+                    }
+                    builder.endObject();
+                }
+                builder.endObject();
+            }
+            builder.endObject();
+            return builder;
+        } catch (IOException e) {
+            throw new UncheckedIOException("Failed to build mappings for " + SYNONYMS_INDEX_CONCRETE_NAME, e);
+        }
+    }
+
+    public void putSynonymsset(String resourceName, SynonymsSet synonymsset, ActionListener<PutSynonymsAction.Response> listener) {
+
+        // TODO Add synonym rules validation
+
+        // Delete synonyms set if it existed previously. Avoid catching an index not found error by ignoring unavailable indices
```

Comment:
PUT should create the element if it does not exist, or replace the element with the new definition in case it already exists ([here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PUT))

Think of the doc index API: `PUT /test-index/_doc/doc-id` would create or update the doc `doc-id`, and not error out in case it already exists (unless `op_type=create` is specified)

That is the reason why documents are deleted - in case rules exist, they need to be replaced by the new ones.

There will be individual rule updates as well:

PUT /_synonyms/<rule_set_id>/<rule_id>

so we will provide the way to update individual rules (or add new ones) as well. Please see [Individual Synonym Rule Update](https://docs.google.com/document/d/1ZIJER4qEab-OacZuPMYRP57g-T45ZkuCvqOnJSCBlUs/edit#heading=h.79pldhtzq9oq) in the Synonym Management API doc.